design-proposals/vm-maintenance-mode.md

```diff
@@ -0,0 +1,61 @@
+# Overview
+Add a new "Maintenance" phase for the VMI/VM that represents a special state where the VM doesn't run and we can safely access the VM disks.
+
+![VMI phase with Maintenance](vm-maintenance-mode-images/VMI_phase.png)
+![VM phase with Maintenance](vm-maintenance-mode-images/VM_phase.png)
+
+The "maintenance" pods are pods created to perform disk operations (like for example to modify the disk with libguestfs-tools). While the VM is in this state, we can safely create maintenance pods that can access VM disks. The maintenance pods must have the same scheduling constraints as the VM in oder to not alterate the scheduling of the VM.
+
+The new Maintenance state differs from Stopped because in this state we refuse to start the VM. Leaving the Maintenance state needs to be an explicit action. We can move the VM/VMI from the Maintenance state back to its previous state only in the case there are not active maintenance pods.
+
+## Motivation
+The main motivation for this proposal is to put the VM in this state in which we can safely perform disk operations. For example, for interactive operations (e.g virtctl guestfs) or batch processing jobs (e.g tekton task). We can access the volumes used by the VM from its specification and attach these volumes to the libguestfs-tools for disk inspection and manipulation. If the VM/VMI is in this state it cannot be in Running state. 
+
+## Goals
+The user can put the VM in this mode and safely access disks without the danger of corrupting the data.
+
+## Non Goals
+  + This proposal does not include any shared storage locking (such as Sanlock).  It is opt-in locking of the VM workload.
+  + This proposal doesn’t protect from pods that directly use the PVC and are not controlled by a VMI or their controller do not verify the state of the VM/VMI
+
+## User Stories
+  + As a user creating VMs, I want to be able to safely access the disks attached to the VM for disk inspection and manipulation with `virtctl guestfs` command
+  + As storage administrator, I want to be able to create periodic or automatic batch processing job on disks of a VM and safely access the disks
+    + Sparsify disks
+  + As VM administrator, I want to be able to create automatic batch processing job on disks of a VM and safely access the disks
+    + Restoring passwords
+    + Creating new users
+    + Installing missing packages
```

Comment:
Might want to mention retrieval of logs (although I am not sure if the guestfs pod has any utilities which would allow for easy export of data.

Reply:
@aglitke do you mean the logs of an application that is running inside the guest?
There is the tool [virt-copy-out](https://libguestfs.org/virt-copy-out.1.html) to copy files or directories outside the disk image. We could attach a special pvc to the maintenance pod where we could copy the logs. For interactive debugging, either with a volume or maybe combining an emptydir with `kubectl cp`?